Codechange: Introduce and use EnumBitSet.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Technically enums have a limited range of values, and using values not listed could cause problems. This affects enums used as bitfields.

Additionally there is sometimes some confusion over whether an enum is used a bit numbers or as bit masks.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Introduce `EnumBitSet` which implements a type-safe wrapper around enum values.

This stores the values of an enum as a bitmask using a defined type. Methods are provided to set/clear/toggle and test bits (vaguely modelled on the `std::bitset<>` methods.

`DECLARE_ENUM_AS_BIT_SET` is no longer needed.

This change converts some existing `enum class` types used as bitmasks to `EnumBitSet`. There are many more that can be done.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

Pre-combined bitmask values are a bit clumsier.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/core/enum_type.hpp:
```diff
@@ -75,4 +75,56 @@ debug_inline constexpr void ToggleFlag(T &x, const T y)
+
+	constexpr auto operator <=>(const EnumBitSet &) const noexcept = default;
+
+	inline constexpr EnumBitSet &set(Tenum value)
```
Shouldn't function names be capitalized?

PeterN replied:
Possibly. I went with lowercase as that matches `std::bitset<>`'s `set(x)` and `test(x)`

~~Although to actually use the same style, I should make `clear(x)` be `reset(x)` and `toggle(x)` be `flip(x)`.~~

(`std::bitset` itself has usuability issues which makes it a pain to use, and the minimum storage size there is `size_t`)

PeterN replied:
Just updated to capitalize the function names.

rubidium42 commented on src/core/enum_type.hpp:
```diff
@@ -75,4 +75,56 @@ debug_inline constexpr void ToggleFlag(T &x, const T y)
+		return (this->data & (1U << to_underlying(value))) != 0;
+	}
+
+	inline constexpr EnumBitSet operator |(Tenum value) const
```
Maybe a bit of documentation here that it is an atypical usage of `|`, i.e. that not the literal value is taken but that it's made the appropriate bitmask first.

Alternatively, won't Flags(Flag::First).set(Flag::Second).set(Flag::Third) work equally well and be less confusing?

PeterN replied:
I've changed this now to a more conventional `inline constexpr EnumBitSet operator |(const EnumBitSet &other) const` which allows combining two EnumBitSets. Due to implicit conversion the original still works, perhaps a bit magic still.

rubidium42 commented on src/core/enum_type.hpp:
```diff
@@ -75,4 +75,56 @@ debug_inline constexpr void ToggleFlag(T &x, const T y)
+		return other.set(value);
+	}
+
+	inline constexpr Tstorage base() const noexcept
```
I reckon the endian writer couldn't be friended, right?

PeterN replied:
Not sure, but having an explicit method to get to the underlying storage is similar to what we have with StrongType.

rubidium42 replied:
True. It was that the only reason I could think of you would need access to the base would be such serialisation code. Anything else shouldn't need access to it, whereas that's quite different for the strong types.

I'm fine with it the way it is right now. It's just a "might be worth it, if it's simple" to prevent misuse.

PeterN replied:
Also I went a bit further to see where this leads ("I have a patch for that..."), and .base() is needed elsewhere, e.g. for values that get reported to NewGRF or to SetDParam()

rubidium42 reviewed (commented):
It might indeed be a good idea to make `enum`s actual enumerations given the associated potential for undefined behaviour. Though their typing's so useful as well.

Removing those nasty macros definitely is a good goal.

rubidium42 commented on src/core/enum_type.hpp:
```diff
@@ -75,4 +75,61 @@ debug_inline constexpr void ToggleFlag(T &x, const T y)
+	using storage_type = Tstorage;
+
+	constexpr EnumBitSet() : data(0) {}
+	constexpr EnumBitSet(Tenum other) : data(to_underlying(other)) {}
```
```suggestion
	constexpr EnumBitSet(Tenum other) : data(1U << to_underlying(other)) {}
```
This should mimic `set`, right? Or maybe just call `set` in the constructor.

Or maybe the following, so you can add an arbitrary number of enums to the constructor.
```
	template <typename... T>
	constexpr EnumBitSet(T... others) : data(0)
	{
		(this->set(others), ...);
	}
```

PeterN committed:
Add: EnumBitSet class to safely use enum class values as a bitset.

This maintains type-safe enum values and type-safe masks, and avoids using undefined enum values.

PeterN committed:
Codechange: Allow using EnumBitSet over network commands.

PeterN committed:
Codechange: Use EnumBitSet for EngineDisplayFlags.

PeterN committed:
Codechange: Use EnumBitSet for TileFlags.

PeterN committed:
Codechange: Use EnumBitSet for ViewportStringFlags.

PeterN committed:
Codechange: Use EnumBitSet for AspectFlags.

PeterN committed:
Codechange: Use EnumBitSet for ExtraEngineFlags.

PeterN committed:
Codechange: Use EnumBitSet for GroupFlags.

PeterN committed:
Codechange: Use EnumBitSet for Border(Flag)s.

PeterN committed:
Codechange: Use EnumBitSet for DepotCommand(Flag)s.
